### ISSUE TITLE
AOTI doesn't error if dynamic shape gets specialized during lowering

### ISSUE THREAD

**Issue by angelayi** [pytorch/pytorch#150364]
### 🐛 Describe the bug

In some passes inductor is able to further shape specialize a dynamic shape. However AOTI does not error on this (during lowering or during runtime). Example:

```python
    def test_aoti_specialization(self):
        from torch._inductor.pattern_matcher import (
            fwd_only,
            PatternMatcherPass,
            register_replacement,
        )
        import torch

        def pattern1(x) -> torch.Tensor:
            return x.sin()
        def replacement1(x) -> torch.Tensor:
            return x.cos()
        patterns = PatternMatcherPass()
        inputs = [torch.randn(6, 5),]
        register_replacement(pattern1, replacement1, inputs, fwd_only, patterns)

        def custom_pass(graph: torch.fx.Graph):
            nodes = [n for n in graph.nodes]
            if nodes[0].meta['val'].shape[0] > 5:
                # replace sin with cos
                patterns.apply(graph)

        torch._inductor.config.post_grad_custom_post_pass = custom_pass

        class M(torch.nn.Module):
            def forward(self, x):
                return x.sin()
        
        ep = torch.export.export(M(), (torch.randn(6, 5),), dynamic_shapes=({0: Dim.DYNAMIC},))
        print(ep)
        path = torch._inductor.aoti_compile_and_package(ep)
        compiled_m = torch._inductor.aoti_load_package(path)
        inp = torch.randn(6, 5)
        self.assertTrue(torch.allclose(compiled_m(inp), inp.cos()))
        inp = torch.randn(3, 5)
        self.assertTrue(torch.allclose(compiled_m(inp), inp.sin()))  # this fails
```

```
===== BEFORE PRE GRAD =====
class GraphModule(torch.nn.Module):
    def forward(self, x: "f32[s35, 5]"):
         # File: /data/users/angelayi/pytorch/moo.py:347 in forward, code: return x.sin()
        sin: "f32[s35, 5]" = torch.ops.aten.sin.default(x);  x = None
        return (sin,)

{s35: VR[2, int_oo]}

===== AFTER POST GRAD =====
class <lambda>(torch.nn.Module):
    def forward(self):
        arg0_1: "f32[s35, 5]"; 
    
        arg0_1, = fx_pytree.tree_flatten_spec([], self._in_spec)
        # No stacktrace found for following nodes
        cos_default: "f32[s35, 5]" = torch.ops.aten.cos.default(arg0_1);  arg0_1 = None
        return (cos_default,)

{s35: VR[6, int_oo]}
```

Maybe we should error while lowering if the shape changes?

Strangely I tried `AOT_INDUCTOR_DEBUG_COMPILE=1` but that didn't seem to actually error when running it, even though the [cpp code](https://www.internalfb.com/phabricator/paste/view/P1771873612?lines=478-484) has the correct check (maybe I did something wrong):
```cpp
    if (arg0_1_size[0] < 6) {
        std::stringstream ss;
        ss << "input_handles[0]: dim value is too small at 0, "
           << "expected it to be >= 6, " << "but got: "
           << arg0_1_size[0] << "\n";
        throw std::runtime_error(ss.str());
    }
```

cc @desertfire @yushangdi @zou3519 

### Versions

main

**Comment by yushangdi** [comment]
To check input shapes, you should use `"AOTI_RUNTIME_CHECK_INPUTS"` instead of `AOT_INDUCTOR_DEBUG_COMPILE=1`.  I verified that you'll see `Error: input_handles[0]: dim value is too small at 0, expected it to be >= 6, but got: 3` if you use `AOTI_RUNTIME_CHECK_INPUTS=1` at run time

Maybe we can close?
